### PR TITLE
Add Rust Module [Rebase & FF]

### PR DESCRIPTION
**Pull request by makubacki**
## Description

Makes some adjustments to allow edk2 Rust-based modules to build
in the repo and adds a hello world Rust DXE module.

These are based on the Rust code from @joschock.

Summary of changes:

---

.pytool/CISettings.py: Set rust-ci scope and use in tree BaseTools

The Rust BaseTool changes are not in the edk2-basetools PIP module,
so this change updates CISettings.py to use the in tree tools by
default.

The `rust-ci` scope is added as an active scope so Rust related
plugins like `RustHostUnitTestPlugin` will run.

---

MsCorePkg: Add RustBootServicesAllocatorDxe

Adds a library crate that implements a global allocator based on
`AllocatePool()`. Memory is allocated from the
`EFI_BOOT_SERVICES_DATA` pool.

---

MsCorePkg: Add HelloWorldRustDxe

Adds a simple Rust based DXE driver to demonstrate how to structure
a DXE driver that only has Rust crate dependencies.

Within the firmware build framework, this is considered a "pure Rust"
DXE driver in that it only has a `Cargo.toml` file in the `[Sources]`
section of the INF with no EDK II library dependencies.

The module uses the `RustAdvancedLoggerDxe` crate (which is a wrapper
around the Advanced Logger protocol) to write debug messages and the
`RustBootServicesAllocatorDxe`.

---

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [x] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

- MsCorePkg CI build
- Boot and test HelloWorldRustDxe on QEMU Q35

## Integration Instructions

See [Rust Build](https://github.com/microsoft/mu_basecore/blob/release/202302/Docs/rust_build.md) for more details about building and using Rust modules.

### CODE REVIEW

**Comment by makubacki** [comment]
This one should go in first: https://github.com/microsoft/mu_basecore/pull/541

Will be picked up right after since this repo build against the tip of mu_basecore release/202302.

**Review by joschock — APPROVED** [review]

File: .pytool/CISettings.py [hunk 1]
```diff
@@ -51,10 +51,12 @@ def AddCommandLineOptions(self, parserObj):
 
     def RetrieveCommandLineOptions(self, args):
         super().RetrieveCommandLineOptions(args)
+
+        # In tree BaseTools are required for Rust build support so enable
+        # it by default.
+        self.UseBuiltInBaseTools = False
         if args.force_piptools:
             self.UseBuiltInBaseTools = True
-        if args.no_piptools:
```
Comment (Javagedes) [inline .pytool/CISettings.py:56]: you should remove the -no_piptools argument if you are no longer checking for it

**Review by Javagedes — APPROVED** [review]

File: .pytool/CISettings.py (hunk 1 above)
Comment (makubacki) [reply .pytool/CISettings.py:56]: I removed it. Just left an option to force pip tools.

**Review by apop5 — APPROVED** [review]

**Comment by makubacki** [comment]
I'm going to have to make some small adjustments to the CodeQL GitHub workflow. Testing that over in [CodeQL + Rust Test PR](https://github.com/microsoft/mu_plus/pull/303) and then I'll push this PR again when done.

**Comment by makubacki** [comment]
> I'm going to have to make some small adjustments to the CodeQL GitHub workflow. Testing that over in [CodeQL + Rust Test PR](https://github.com/microsoft/mu_plus/pull/303) and then I'll push this PR again when done.

Change for that is ready over in [.sync/codeql: Add cargo make and install rust-src ](https://github.com/microsoft/mu_devops/pull/235). I'll catch it in the pending file sync PRs across the repos before they're completed.

**Comment by makubacki** [comment]
> Change for that is ready over in [.sync/codeql: Add cargo make and install rust-src ](https://github.com/microsoft/mu_devops/pull/235). I'll catch it in the pending file sync PRs across the repos before they're completed.

Sync PR with the CodeQL workflow change is merged now.

**Pull request merged**